### PR TITLE
Preserve history when deploying to gh-pages.

### LINKED ISSUE #704: Allow github pages deployments to retain history
The [W3C deployment scripts](https://gist.github.com/domenic/ec8b0fc8ab45f39403dd) that I'm trying to replace retain the history of the gh-pages branch.

If this is a change you'd accept, would you rather an option to retain history, or just unconditionally do it?

### PR DESCRIPTION
**Pull request by jyasskin**
Fixes #704.

This change sits on top of #695, which is mostly reviewed.

@webknjaz, does this look reasonable to you?

Test deploys:
* https://travis-ci.org/jyasskin/testing/builds/304151299 testing a deploy with no existing gh-pages branch.
* https://travis-ci.org/jyasskin/testing/builds/304151521 testing a deploy with an existing gh-pages branch
* https://travis-ci.org/jyasskin/testing/builds/304151699 testing that a no-op deploy is skipped.

You can see the resulting gh-pages branch at https://github.com/jyasskin/testing/commits/gh-pages.

### CODE REVIEW

**Comment by BanzaiMan** [comment]
Thanks for the PR. Could you explain why you are proposing `deploy-key` and `deploy-key-key`? These names don't feel particularly self-revealing, and forcing encryption seems rather strange.

**Comment by jyasskin** [comment]
@BanzaiMan I explained the need for encryption in https://github.com/travis-ci/dpl/pull/695#discussion_r146095804 item 2.

**Comment by BanzaiMan** [comment]
Oh, sorry. If this is dependent on #695, could you add to that PR instead of opening a new one? Thanks.

**Comment by jyasskin** [comment]
They're totally different features; they just have merge conflicts. If you could review #695 with less than a month latency, this wouldn't need to sit on it.

File: lib/dpl/provider/pages.rb [hunk 1]
```diff
@@ -28,7 +30,14 @@ def initialize(context, options)
 
         @gh_fqdn = fqdn
         @gh_url = options[:github_url] || 'github.com'
-        @gh_token = option(:github_token)
+        @gh_token = options[:github_token]
+        @gh_deploy_key = options[:deploy_key]
+        unless @gh_token || @gh_deploy_key
```
Comment (webknjaz) [inline lib/dpl/provider/pages.rb]: This must go to the `check_auth` method

File: lib/dpl/provider/pages.rb [hunk 2]
```diff
@@ -48,29 +57,70 @@ def check_auth
...
+            setup_git_ssh("#{gitdir}/git-ssh", "#{gitdir}/deploykey")
+          end
+          Dir.mktmpdir {|tmpdir|
+            # A failed `git clone` removes its target directory, so use a
+            # subdirectory to make sure we can safely re-create it.
+            tmpdir = "#{tmpdir}/deploy"
+            unless create_git_dir(tmpdir)
+              error "Couldn't create checkout to push to #{@gh_ref}:#{@target_branch}"
+            end
+            context.shell "rsync -r --exclude .git --delete #{@build_dir}/ #{tmpdir}" or error "Could not copy #{@build_dir}."
```
Comment (webknjaz) [inline lib/dpl/provider/pages.rb]: why rsync?

**Comment by webknjaz** [comment]
@jyasskin sorry, I was offline for a while. I'll try reviewing this and #695 in detail soon. Did you backport some commits from https://github.com/webknjaz/dpl/tree/feature/enable-gh-pages-keep-history?

File: lib/dpl/provider/pages.rb [hunk 3]
```diff
@@ -48,29 +57,70 @@ def check_auth
...
+
       def github_deploy
-        context.shell 'rm -rf .git > /dev/null 2>&1'
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
-        context.shell 'git init' or raise 'Could not create new git repo'
-        context.shell "git config user.email '#{@gh_email}'"
-        context.shell "git config user.name '#{@gh_name}'"
         context.shell "echo '#{@gh_fqdn}' > CNAME" if @gh_fqdn
         context.shell 'git add .'
+        if context.shell "git diff --cached --quiet"
```
Comment (webknjaz) [inline lib/dpl/provider/pages.rb]: I'd prefer to give user more control over an empty commit https://github.com/webknjaz/dpl/commit/c8201a93163142dc43245ec5c097d6a054307245, having reasonable default, of course.

**Comment by webknjaz** [comment]
Also, I'd prefer this to be merged before #695 and be decoupled.

File: lib/dpl/provider/pages.rb [hunk 4]
```diff
@@ -86,9 +86,12 @@ def create_git_dir(gitdir)
       end
 
       def github_deploy
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
```
Comment (webknjaz) [inline lib/dpl/provider/pages.rb]: This change is unrelated and must be feature-flagged

**Comment by webknjaz** [comment]
@jyasskin I've created PR #719 to showcase you my old code. Please review it. I feel like it's better in terms of public API

File: lib/dpl/provider/pages.rb (hunk 1 above)
Comment (jyasskin) [reply lib/dpl/provider/pages.rb]: Done in 7f4184d4f19bb2191106f05cbf260a95b78fece3.

File: lib/dpl/provider/pages.rb (hunk 3 above)
Comment (jyasskin) [reply lib/dpl/provider/pages.rb]: That's plausible, but new options have a cost, and my impression is that @BanzaiMan doesn't want this one. I'm happy to add this option if they prefer to have it. See also https://github.com/travis-ci/dpl/pull/719#pullrequestreview-79236861.

File: lib/dpl/provider/pages.rb (hunk 2 above)
Comment (jyasskin) [reply lib/dpl/provider/pages.rb]: It simplifies the process of removing the source .git file and the destination files that aren't present in the source. More discussion in https://github.com/travis-ci/dpl/pull/719#discussion_r153264248 and https://github.com/travis-ci/dpl/pull/719#discussion_r153267193.

**Comment by 153957** [comment]
As I mentioned in #719:
> I am in favor if keeping the option to force push (discard history), I do not care which way will be the default, but we have some repositories in which the changes in the gh-pages branch are quite large (binary files) and it is not important to keep the history, so to keep the repo size down discarding history is a great feature.

**Comment by BanzaiMan** [comment]
This can now be closed, since #719 has been merged, right?

**Comment by webknjaz** [comment]
Correct.
